src/rdkafka_broker.c

```diff
@@ -4206,73 +4208,113 @@ static int rd_kafka_broker_thread_main (void *arg) {
                                 rd_kafka_broker_serve(rkb, 1000);
 
                         /* Throttle & jitter reconnects to avoid
-                         * thundering horde of reconnecting clients after
-                         * a broker / network outage. Issue #403 */
+                        * thundering horde of reconnecting clients after
+                        * a broker / network outage. Issue #403 */
                         backoff = rd_kafka_broker_reconnect_backoff(rkb,
-                                                                    rd_clock());
+                                                                rd_clock());
                         if (backoff > 0) {
                                 rd_rkb_dbg(rkb, BROKER, "RECONNECT",
-                                           "Delaying next reconnect by %dms",
-                                           backoff);
+                                        "Delaying next reconnect by %dms",
+                                        backoff);
                                 rd_kafka_broker_serve(rkb, (int)backoff);
                                 continue;
                         }
 
-			/* Initiate asynchronous connection attempt.
-			 * Only the host lookup is blocking here. */
-			if (rd_kafka_broker_connect(rkb) == -1) {
-				/* Immediate failure, most likely host
-				 * resolving failed.
-				 * Try the next resolve result until we've
-				 * tried them all, in which case we sleep a
-				 * short while to avoid busy looping. */
-				if (!rkb->rkb_rsal ||
-                                    rkb->rkb_rsal->rsal_cnt == 0 ||
-                                    rkb->rkb_rsal->rsal_curr + 1 ==
-                                    rkb->rkb_rsal->rsal_cnt)
+                        /*
+                        * SASL/OAUTHBEARER is unable to connect unless a valid token is
+                        * available, and a valid token CANNOT be available unless/until
+                        * an initial token retrieval succeeds, so wait for this
+                        * precondition if necessary.
+                        */
+                        if (rkb->rkb_rk->rk_oauthbearer) {
+                                int has_token = 0;
+                                int est_total_wait_ms = 0;
+                                int max_wait_ms = 5000;
```

Comment:
Should this be configurable, or a documented constant? I imagine making this value more visible might help the user in the case where authentication fails because the callback is slow to provide a token.

Reply:
Good point.  I used 5 seconds because [that is what kafkacat uses as a timeout when making a connection](https://github.com/edenhill/kafkacat/blob/master/kafkacat.c#L669), and I figured there is no need to wait longer than what the client is going to wait, but different clients may behave differently. @We could document a hard-coded value or create a new config (such as `sasl.oauthbearer.initial.timeout.ms`).  Note that this is not something that has an impact during the whole life of the client -- it is just a timeout for the initial token refresh.  It feels to me that hard-coding a large value (maybe 30 seconds instead of the 5 seconds I chose) and documenting that value clearly is probably reasonable.  @mkaminski1988 Do you agree?

Reply:
+1 Another option could be to append the timeout value in the debug messaging, something like `Main broker thread will exit due to no initial token after %d seconds`

Reply:
👍 Increased it to 30 seconds and added it to debug messages